### ISSUE TITLE
Importing/Forking Concepts & Codelist review

### ISSUE THREAD

**Issue by JackScanlon** [SwanseaUniversityMedical/concept-library#1096]
# Closure of issue
Closed as task was split into separate issues:
- #1152 
- #1153 

# Issue
### Details
> **[!] Note:** Blocked by #1136 

This is a continuation of:
- #1076 

### Requirements
The following needs to be discussed further before implementation:
1. How we expect to design the concept selection method within the phenotype builder
    - How do we avoid usage of modals?
    - How do we make this accessible?
    - Do we only show published concepts? If so, do users find them through searching for children of Phenotypes?

2.  How do we intend to implement concept imports?
    - If a user imports a concept, can it be modified?
    - If we allow forking, how are would we like to handle this so that the root concept is linked?
    - If concept codelists can be imported as a ruleset, is it only the final codelist that can be imported, or should we allow importing of single rulesets?
    - If we allow users to import single rulesets, how do we handle forking so that the original creator is credited and/or linked to that concept?

3. How do we intend the codelist review system to work, e.g.:
    - Is it a requirement for a codelist to be reviewed before publication? 
    - Do we need to have a state enum so that users can review a codelist multiple times until it's in an 'accepted' state?
    - If we intend for the review section to be within both the phenotype builder, and it's own design, how do we expect to design this to fit within the phenotype builder?

# Considerations
Change of Components→Rulesets, but consideration of the following tasks need to be made for importing legacy Concept:
- #362

# Tasks

### Changes to Phenotype Creation
- [ ] Changes to creation process
    - [ ] Ability to import top-level Concepts
    - [ ] Ability to import final Concept codelists as Rulesets
- [ ] Changes to backend
    - [ ] Source type
    - [ ] Tracking forking

### Reviewing Entities
- [ ] Codelist review frontend
- [ ] Codelist review backend
    - [ ] Update current save / publication method to incl. review status

**Comment by DSThayer** [comment]
1) I'm starting to wonder if we are reaching the level of complexity where we really need a second screen. I was already wondering that given that it wasn't feasible to include the review step in the main screen. Perhaps creating/editing a concept should go back to being a separate screen, albeit one integrated with the phenotype workflow, rather than standalone? 

This would give more space for all of the things that have to go in there, like rule building. If we did this, I would also like it to have a wide table that is not constrained by the typical page content width, so that one can easily see the table and all the rules. Something to think about, discuss further? 

2) Basically a rule can be added which is pointing to the concept in another phenotype. I think behind the scenes we actually import the codes so everything is one place, but to the user it looks more like a pointer to the content than an "import".

This feature exists in concepts already, and we want to exactly replicate that functionality (just with a better UI).

It is always the final codes of a concept; the rules inside of it are not visible in this context.

It cannot be modified. 

The reference to the original content is always maintained (just like we save the regex if you ran a search). So if you fork a phenotype or whatever, this would come along to the new one. 

It is a reference to a specific version of the Concept (I don't think concepts will be independently versioned now, so it would be a phenotype version?). It won't change even if someone modifies the concept in its original setting.

Complicated part: if the original concept _was_ updated, we want some way to highlight that it's an older version and let the user update their own phenotype to the latest? This is a "not version 1" feature that needs more discussion.

3) The whole codelist review functionality is optional.

The codelist review is an optional feature that can be added to a concept or not. (Perhaps implemented as a type of exclusion rule). 

Maybe saying "you have unreviewed codes, are you sure you want to publish?" would be a good idea if a codelist review exists within a phenotype and is incomplete...I wouldn't block publication, though. Also highlighting that a review is incomplete and action is needed on the "My content" page would be great.

Each individual code has a trimodal state, "include/exclude/unreviewed". The review is complete if there are no "unreviewed" codes in the list. (It can become incomplete again if you modify the code list and therefore have unreviewed codes again).

Review decisions on codes are saved forever in the concept. If you mess with your search term inclusion rule, accidentally deleting all the codes from your concept, then bring them back, it should remember the review decisions that were previously made on those codes. (I think we discussed it in terms of storing the "include" or "exclude" decisions with codes in a table, and then "unreviewed" is implemented by codes not being present in that table)
